File: services/opencode_session.py
"""Best-effort conversation affinity for clients without session headers."""

import hashlib
import hmac
import json
import uuid
from collections.abc import Mapping

from services.client_headers import with_client_defaults

MAX_SESSION_BODY_BYTES = 1024 * 1024


def _conversation_anchor(payload):
    if not isinstance(payload, dict):
        return None
    metadata = payload.get("metadata")
    metadata = metadata if isinstance(metadata, dict) else {}
    conversation = payload.get("conversation")
    for value in (
        payload.get("session_id"),
        payload.get("conversation_id"),
        metadata.get("session_id"),
        metadata.get("conversation_id"),
        conversation.get("id") if isinstance(conversation, dict) else conversation,
    ):
        if isinstance(value, str) and value.strip():
            return ["conversation", value]

    messages = payload.get("messages", payload.get("input"))
    if isinstance(messages, str):
        messages = [{"role": "user", "content": messages}]
    if not isinstance(messages, list):
        return None
    opening = []
    for message in messages:
        if not isinstance(message, dict):
            continue
        role, content = message.get("role"), message.get("content")
        if role not in ("system", "developer", "assistant", "user"):
            continue
        if not isinstance(content, (str, list)) or not content:
            continue
        opening.append([role, content])
        if role == "user":
            return [
                "opening",
                payload.get("system"),
                payload.get("instructions"),
                opening,
            ]
    return None


def with_opencode_request_session(
    headers: Mapping[str, str], data: bytes | None
) -> dict[str, str]:
    """Preserve real IDs; otherwise infer affinity without storing prompts or keys."""
    result = with_client_defaults(headers, "opencode")
    if result.get("X-Opencode-Session"):
        return result

    lower = {name.lower(): value for name, value in headers.items()}
    scope = lower.get("authorization") or lower.get("x-api-key") or ""
    anchor = None
    if scope and data and len(data) <= MAX_SESSION_BODY_BYTES:
        try:
            anchor = _conversation_anchor(json.loads(data))
            encoded = json.dumps(
                anchor,
                sort_keys=True,
                ensure_ascii=False,
                separators=(",", ":"),
                allow_nan=False,
            ).encode("utf-8")
        except (ValueError, UnicodeError, RecursionError):
            anchor = None
    if anchor is not None:
        digest = hmac.new(scope.encode("utf-8"), encoded, hashlib.sha256).hexdigest()
        session = f"multillm_v1_{digest}"
    else:
        # No conversation evidence: isolate this logical request, not the fleet.
        session = f"multillm_request_{uuid.uuid4()}"
    result["X-Opencode-Session"] = session
    return result

File: worker/opencode-session.mjs
import { withOpencodeSession } from "./client-headers.mjs";

export const MAX_SESSION_BODY_BYTES = 1024 * 1024;

function conversationAnchor(payload) {
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) return null;
  for (const value of [
    payload.session_id, payload.conversation_id,
    payload.metadata?.session_id, payload.metadata?.conversation_id,
    typeof payload.conversation === "object" ? payload.conversation?.id : payload.conversation,
  ]) {
    if (typeof value === "string" && value.trim()) return ["conversation", value];
  }
  let messages = payload.messages ?? payload.input;
  if (typeof messages === "string") messages = [{ role: "user", content: messages }];
  if (!Array.isArray(messages)) return null;
  const opening = [];
  for (const message of messages) {
    const { role, content } = message ?? {};
    if (!["system", "developer", "assistant", "user"].includes(role)) continue;
    if (!(typeof content === "string" || Array.isArray(content)) || !content.length) continue;
    opening.push([role, content]);
    if (role === "user") return ["opening", payload.system ?? null, payload.instructions ?? null, opening];
  }
  return null;
}

function canonical(value) {
  if (Array.isArray(value)) return value.map(canonical);
  if (value && typeof value === "object") {
    return Object.fromEntries(Object.keys(value).sort().map((key) => [key, canonical(value[key])]));
  }
  return value;
}

async function boundedPayload(request) {
  const declaredLength = request.headers.get("Content-Length");
  const contentType = request.headers.get("Content-Type")?.split(";", 1)[0].trim().toLowerCase();
  // Unknown-length uploads must start streaming immediately, without a body probe.
  if (!request.body || !declaredLength || !/^\d+$/.test(declaredLength)
      || Number(declaredLength) > MAX_SESSION_BODY_BYTES || contentType !== "application/json") return null;
  const reader = request.clone().body.getReader();
  // Cloned streams can stall. Affinity discovery must not delay raw transport indefinitely.
  const timer = setTimeout(() => { void reader.cancel().catch(() => {}); }, 1000);
  const chunks = [];
  let length = 0;
  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      length += value.byteLength;
      if (length > MAX_SESSION_BODY_BYTES) return null;
      chunks.push(value);
    }
    const bytes = new Uint8Array(length);
    let offset = 0;
    for (const chunk of chunks) { bytes.set(chunk, offset); offset += chunk.byteLength; }
    return JSON.parse(new TextDecoder("utf-8", { fatal: true }).decode(bytes));
  } catch {
    return null;
  } finally {
    clearTimeout(timer);
    void reader.cancel().catch(() => {});
  }
}

export async function withOpencodeRequestSession(request, source) {
  const headers = withOpencodeSession(source);
  if (headers.has("x-opencode-session")) return headers;
  const scope = headers.get("Authorization") || headers.get("X-Api-Key");
  let anchor = null;
  try {
    if (scope) anchor = canonical(conversationAnchor(await boundedPayload(request)));
  } catch {
    // Invalid or excessively nested input remains the provider's validation responsibility.
  }
  let session = `multillm_request_${crypto.randomUUID()}`;
  if (anchor !== null) {
    const encoder = new TextEncoder();
    const key = await crypto.subtle.importKey("raw", encoder.encode(scope), { name: "HMAC", hash: "SHA-256" }, false, ["sign"]);
    const digest = new Uint8Array(await crypto.subtle.sign("HMAC", key, encoder.encode(JSON.stringify(anchor))));
    session = `multillm_v1_${Array.from(digest, (byte) => byte.toString(16).padStart(2, "0")).join("")}`;
  }
  headers.set("x-opencode-session", session);
  return headers;
}
